# 一个异步的例子
首先，按照 [入门指南](https://facebook.github.io/jest/docs/getting-started.html#using-babel) 中的说明，在 Jest 中启用 Babel 支持。

让我们实现一个从 API 获取用户数据并返回 user name 的简单模块。

```javascript
// user.js
import request from './request';

export function getUserName(userID) {
  return request('/users/' + userID).then(user => user.name);
}
```

在上面的实现中，我们期望 request.js 模块返回一个 promise。我们链式调用 then 来接收 user name。

现在想象一个访问网络并获取一些用户数据的 request.js 的实现。

```javascript
// request.js
const http = require('http');

export default function request(url) {
  return new Promise(resolve => {
    // This is an example of an http request, for example to fetch
    // user data from an API.
    // This module is being mocked in __mocks__/request.js
    http.get({path: url}, response => {
      let data = '';
      response.on('data', _data => data += _data);
      response.on('end', () => resolve(data));
    });
  });
}
```

因为我们不想在我们的测试中去访问网络，我们将在 \_\_mocks\_\_ 文件夹中创建一个手动模拟的 request.js 模块。它可能看起来像这样：

```javascript
// __mocks__/request.js
const users = {
  4: {name: 'Mark'},
  5: {name: 'Paul'},
};

export default function request(url) {
  return new Promise((resolve, reject) => {
    const userID = parseInt(url.substr('/users/'.length), 10);
    process.nextTick(
      () => users[userID] ? resolve(users[userID]) : reject({
        error: 'User with ' + userID + ' not found.',
      })
    );
  });
}
```

现在我们来为我们的异步功能编写一个测试。

```javascript
// __tests__/user-test.js
jest.mock('../request');

import * as user from '../user';

// The assertion for a promise must be returned.
it('works with promises', () => {
  expect.assertions(1);
  return user.getUserName(4).then(data => expect(data).toEqual('Mark'));
});
```

我们调用 jest.mock('../request') 告诉 Jest 使用我们的手动模拟。it 期望返回值是一个将要 resolved 的 Promise。您可以链式调用多个 Promise，随时调用 expect，只要你在最后返回 Promise。

### .resolves
*Jest 20.0.0+ 支持*

有一种简洁的方法是使用 resolves 来获取 promise 成功的值，使它可以连接任何匹配器。如果 promise 失败了，断言将失败。

```javascript
it('works with resolves', () => {
  expect.assertions(1);
  return expect(user.getUserName(5)).resolves.toEqual('Paul');
});
```

### async/await
使用 async/await 语法编写测试很容易。前面的例子可以这样写：

```javascript
// async/await 可以这样用.
it('works with async/await', async () => {
  expect.assertions(1);
  const data = await user.getUserName(4);
  expect(data).toEqual('Mark');
});

// async/await 也可以和 .resolves 一起用
it('works with async/await and resolves', async () => {
  expect.assertions(1);
  await expect(user.getUserName(5)).resolves.toEqual('Paul');
});
```

要在项目中使用 async/await，请安装 babel-preset-env 并在 .babelrc 文件中进行设置。

### 错误处理

可以使用 .catch 方法处理错误。确保添加 expect.assertions 进行验证，它可以保证一定数量的断言被调用。否则一个 promise 成功了不会让测试失败。

```javascript
// 使用 Promise.catch 来测试异步错误
test('tests error with promises', async () => {
  expect.assertions(1);
  return user.getUserName(2).catch(e =>
    expect(e).toEqual({
      error: 'User with 2 not found.',
    })
  );
});

// 或者使用 async/await
it('tests error with async/await', async () => {
  expect.assertions(1);
  try {
    await user.getUserName(1);
  } catch (e) {
    expect(e).toEqual({
      error: 'User with 1 not found.',
    });
  }
});
```

### .rejects
*Jest 20.0.0+ 支持*

.rejects 的工作方式就像 .resolves 一样。如果 promise 成功，测试将自动失败。

```javascript
// 使用 .rejects 测试异步错误
it('tests error with rejects', () => {
  expect.assertions(1);
  return expect(user.getUserName(3)).rejects.toEqual({
    error: 'User with 3 not found.',
  });
});

// 或者使用 async/await 和 .rejects
it('tests error with async/await and rejects', async () => {
  expect.assertions(1);
  await expect(user.getUserName(3)).rejects.toEqual({
    error: 'User with 3 not found.',
  });
});
```

此示例的代码在 [examples/async](https://github.com/facebook/jest/tree/master/examples/async)






































*
